src/ctap/data_formats.rs

```diff
@@ -445,66 +445,93 @@ pub struct PublicKeyCredentialSource {
     pub user_handle: Vec<u8>, // not optional, but nullable
     pub other_ui: Option<String>,
     pub cred_random: Option<Vec<u8>>,
+
+    /// Contains the unknown fields when parsing a CBOR value.
+    ///
+    /// Those fields could either be deleted fields from older versions (they should have reserved
+    /// tags) or fields from newer versions (the tags should not be reserved). If this is empty,
+    /// then the parsed credential is probably from the same version (but not necessarily).
+    pub unknown_fields: BTreeMap<cbor::KeyType, cbor::Value>,
+}
+
+// We serialize credentials for the persistent storage using CBOR maps. Each field of a credential
+// is associated with a unique tag, implemented with a CBOR unsigned key.
+#[repr(u64)]
```

Comment:
I don't see the use case of specifying the representation here. CBOR can dynamically represent integers of up to 64 bits via varint encoding. And once parsed from CBOR into an enum, there's no need to reserve 64 bits of memory. It's unlikely that there will ever be 2^32 variants in this enum, so forcing 64 bits is wasting space (and our typical target CPU architecture has words of 32 bits).

Reply:
We could add smaller integer types to `KeyType`. Our CBOR implementation is efficient in output, but not necessarily in memory. I was wondering if your `repr` statement addresses something unique to your storage. If not, it adds more confusion than documentation, at least for me.

Reply:
The current CBOR implementation seems to be "canonical" in the sense that there is a bijection between the binary data and the Rust structure (modulo map ordering I guess). Adding smaller integer types to `KeyType` would break this property and also would not fix any memory issue. The problem is that `KeyType` shares variant of different sizes. One solution would be to use one level of indirection (like in functional languages), but this comes with the price of an indirection so there's a compromise to be done. I don't think it's worth it. If there are memory issues, they should be analyzed and the bigger problems should be addressed first.

Reply:
Agreed we should change the CBOR implementation. Since the line is gone in in your current commit, I consider this resolved.

Reply:
Yes, I removed the `repr` because it was meant for documentation and apparently was confusing, so it was obviously not a good documentation.